### ISSUE TITLE
User Defined Function Responses with content larger than >6MB, Spill support for UDFs?

### ISSUE THREAD

**Issue by rustyconover** [awslabs/aws-athena-query-federation#764]
Hi! thanks for creating this repo it's full of great examples of federated queries.

I've been implementing my own handlers for federated queries using Python and PyArrow. 

Specifically, I've found it challenging to find documentation on how to handle the case where a user-defined function wants to return more than 6MB in data to Athena.  The reason why the 6MB size limit is important is that it is the largest response that Lambda can return to a caller.

I haven't found any examples in the test code of this repo that show how larger UDF responses can be handled.  Normally my responses are in the format:

```json
{
  "@type": "UserDefinedFunctionResponse",
  "records": {
    "aId": str(uuid4()),
    "schema": output_schema,
    "records": output_records
  },
  "methodName": event["methodName"]
}
```

Those responses work great when there is less than 6 MB of data to return.

But I've tried responses like:

```json
{
  "@type": "RemoteUserDefinedFunctionResponse",
  "encryptionKey": None,
  "schema": output_schema,
  "remoteBlocks": [
    {
      "@type": "S3SpillLocation",
      "bucket": "[redacted]",
      "key": "athena-spills/spill.1",
      "directory": False,
    },
  ],
  "methodName": event["methodName"]
}
```

There doesn't seem to be any documentation about how to handle this case where Athena User Defined Function wants to spill its results to S3 so that it can return more than 6 MB of data.

Can you please provide documentation or examples of a UDF returning more than 6 MB of response data?

Are all of the API request and response data types documented, and I'm overlooking the spec?

Thank you

Rusty

**Comment by rustyconover** [comment]
This may be related to #450

**Comment by avirtuos** [comment]
Today, UDFs are not allowed to return more than 6MB. This is a hard limit as we do not support spill to S3 for UDFs. Athena will not call with a page size larger than 1MB today (if i recall correctly) so unless you UDF is inflating the data by ~6x for response you should be safe. We have plans to remove the 6Mb restriction but its currently pretty far down the road as it will come with a bunch of other enhancements that we are not yet ready to talk about concretely. 

Can you tell me a bit more about how / why you are running into the 6MB limit? What does your UDF do?

**Comment by avirtuos** [comment]
regarding the related issue you posted, can you confirm what size the input batch was to the UDF? if its more than 1Mb then, yes you might be hitting #450 but my understanding is that this issue was rare, if still present.

**Comment by rustyconover** [comment]
Hi @avirtuos, 

Thanks for writing back.

Right now, my UDF it allows JMESPath expressions to be evaluated against JSON data.  You can see what JMESPath can do here: https://jmespath.org/

Here is an example query:

```
USING EXTERNAL FUNCTION jmespath_search(path varchar, data varchar)
RETURNS varchar
LAMBDA 'jmespath-lambda'
SELECT  jmespath_search('foo', '{ "foo": {"baz": "apple"}}') 
```

JMESPath can select/filter elements out of JSON structures but it can also be used to compose entirely new JSON objects.

I find it easier to use rather than Athena's `json_extract` which can only use a JSONPath expression.

Here is my UDF:

https://gist.github.com/rustyconover/6cca64403f2c1cba8be0ddc3f81f27a1

If you read my UDF's source code, please forgive the hacks to serialize the schema and the record batch separately.  Maybe in the future, the interface of Athena can be changed such that Arrow's IPC format can be supplied in one field rather than split out into the schema and the record batch separately.

Since the input JSON data is unpredictable it is hard to guarantee that the data returned will be smaller than the source data.

**Comment by rustyconover** [comment]
Another excellent example of a useful UDF would list the contents of a Zip file stored at an S3 URI and return an array of varchars for each file contained inside the Zip file.  

That UDF may result in more than a 6x increase in the input size since zip files can contain many files.

**Comment by zxkane** [comment]
> Another excellent example of a useful UDF would list the contents of a Zip file stored at an S3 URI and return an array of varchars for each file contained inside the Zip file.
> 
> That UDF may result in more than a 6x increase in the input size since zip files can contain many files.

Yes. Hopefully it will have a setting to limit the batch size for the UDF functions.
